### PR TITLE
boards: MikroE Clicker 2: Added board

### PR DESCRIPTION
**Pull request by pbeTrifork**
Added MikroE Clicker 2 board.

Discussed with @henrikbrixandersen the best way to cope with the two mikrobus ports on the board, didn't get anything conclusive so added as aliases for now

Signed-off-by: Paul M. Bendixen <pbe@trifork.com>

### CODE REVIEW

File: boards/arm/mikroe_clicker_2/doc/mikroe_clicker_2.rst [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+   $ openocd -f /usr/share/openocd/scripts/interface/stlink-v2.cfg \
+       -f /usr/share/openocd/scripts/target/stm32f4x.cfg -c init\
+       -c "reset halt" -c "stm32f4x unlock 0" -c "reset run" -c shutdown
+
+Here is an example for the :ref:`hello_world` application.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: mikroe_cliker_2
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/doc/mikroe_clicker_2.rst:99]:
```suggestion
   :board: mikroe_clicker_2
```

File: boards/arm/mikroe_clicker_2/doc/mikroe_clicker_2.rst [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+   :goals: build flash
+
+You should see the following message on the console:
+
+.. code-block:: console
+
+   Hello World! mikroe_cliker_2
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/doc/mikroe_clicker_2.rst:106]:
```suggestion
   Hello World! mikroe_clicker_2
```

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		led0 = &ld1;
+		led1 = &ld2;
+		sw0 = &user_button_0;
+		sw1 = &user_button_1;
+		mikrobus-1-i2c = &i2c3;
+		mikrobus-1-spi = &spi3;
+		mikrobus-1-uart = &usart2;
+		mikrobus-2-i2c = &i2c2;
+		mikrobus-2-spi = &spi2;
+		mikrobus-2-uart = &usart3;
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts:57]: These should not be aliases, but node labels. See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/70ff1d73c42f2e174c49b1e217712b52ad6ae78b/boards/arm/lpcxpresso55s16/lpcxpresso55s16_common.dtsi#L138 for an example.

File: boards/arm/mikroe_clicker_2/pinmux.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	{STM32_PIN_PA8, STM32F4_PINMUX_FUNC_PA8_I2C3_SCL},
+	{STM32_PIN_PC9, STM32F4_PINMUX_FUNC_PC9_I2C3_SDA},
+#endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adc1), okay) && CONFIG_ADC
+	{STM32_PIN_PA2, STM32F4_PINMUX_FUNC_PA2_ADC123_IN2},
+	{STM32_PIN_PA3, STM32F4_PINMUX_FUNC_PA3_ADC123_IN3},
+#endif
+};
+
+static int pinmux_black_f407ve_init(struct device *port)
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/pinmux.c:57]: This functions name should reflect the board name.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, a few minor corrections below.

File: boards/arm/mikroe_clicker_2/board.cmake [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=STM32F407VG" "--speed=4000")
+board_runner_args(openocd --cmd-pre-init "source [find interface/jlink.cfg]")
+board_runner_args(openocd --cmd-pre-init "source [find target/stm32f4x.cfg]")
```
[the remark this answers is not in the document]
Comment (pbeSmartDevice) [reply boards/arm/mikroe_clicker_2/board.cmake:5]: to be perfectly honest I don't know, but If I leave either of them out it doesn't

File: boards/arm/mikroe_clicker_2/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_PINMUX)
+zephyr_library()
+zephyr_library_sources(pinmux.c)
+zephyr_library_include_directories(${ZEPHYR_BASE}/drivers)
```
[the remark this answers is not in the document]
Comment (pbeSmartDevice) [reply boards/arm/mikroe_clicker_2/CMakeLists.txt]: Apparently not.

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+&i2c2 {
+	status = "okay";
+};
+
+&i2c3 {
+	status = "okay";
+};
+
+&adc1 {
+	status ="okay";
```
Comment (pbeSmartDevice) [inline boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts:98]: Would it be possible to have labels for the channels here ? I couldn't find (or understand) the documentation of it.
e.g.:

    mikrobus_1_adc: adc1_ch2 {
        channel = <2>;
    };
    mikrobus_2_adc: adc1_ch3 {
        channel = <3>;
    };

File: boards/arm/mikroe_clicker_2/Kconfig.board [hunk 8]
```diff
@@ -0,0 +1,8 @@
+# "MikroE Clicker 2 board configuration
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/Kconfig.board:1]:
```suggestion
# MikroE Clicker 2 board configuration
```

File: boards/arm/mikroe_clicker_2/Kconfig.defconfig [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# "MikroE Clicker 2 board configuration
```
Comment (henrikbrixandersen) [inline boards/arm/mikroe_clicker_2/Kconfig.defconfig:1]:
```suggestion
# MikroE Clicker 2 board configuration
```

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts (hunk 7 above)
Comment (henrikbrixandersen) [reply boards/arm/mikroe_clicker_2/mikroe_clicker_2.dts:98]: That's not currently possible. We would need something along the line of the gpio-nexus for ADCs in order to support this...

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by pbeSmartDevice** [comment]
Ok so it seems the problem in the tests is that I use defconfig to get a uart console on the usb interface, but when running the non-multithreaded kernel test POSIX is not defined and then the USB stack cannot be defined.
Should I move UART_CONFIG_ON_DEV_NAME to Kconfig.defconfig and make it depend on USB or something?
Any ideas on fixing this would be appreciated.

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2_defconfig [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+CONFIG_SERIAL=y
+
+# console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_USB=y
```
Comment (erwango) [inline boards/arm/mikroe_clicker_2/mikroe_clicker_2_defconfig:15]: As documented in https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations, for  portability purpose, it is application responsibility to enable "non basic" features.
As a consequence, USB should not be enabled here, also please check symbols dependencies (for instance CONFIG_USB_DC_STM32 will be set to y if CONFIG_USB=y).
If needed, USB console can be conditionally configured in board.defconfig.

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2_defconfig (hunk 10 above)
Comment (pbeSmartDevice) [reply boards/arm/mikroe_clicker_2/mikroe_clicker_2_defconfig:15]: I couldn't find any examples using board.defconfig, I tried making it work in Kconfig.defconfig, but the dependencies were too much to handle.

File: boards/arm/mikroe_clicker_2/mikroe_clicker_2.yaml [hunk 11]
```diff
@@ -0,0 +1,14 @@
+identifier: mikroe_clicker_2
+name: MikroE Clicker 2 board for STM32
```
[the remark this answers is not in the document]
Comment (pbeSmartDevice) [reply boards/arm/mikroe_clicker_2/mikroe_clicker_2.yaml]: :+1:

**Comment by erwango** [comment]
@pbeSmartDevice Sorry, it seems I missed your PR.
If you're still interesting in getting it merged (which would be nice), can you rebase on top of master and move UART configuration to device tree ?

**Comment by pbeSmartDevice** [comment]
I tried to update it, but I can't seem to find the right include and I can't figure out how to include something that is not a file in my `.dts`

**Comment by erwango** [comment]
> I tried to update it, but I can't seem to find the right include and I can't figure out how to include something that is not a file in my `.dts`

Do you mean this `#include <st/f4/stm32f407v(e-g)tx-pinctrl.dtsi>` ?
If not working, please try to run `west update` and check that the latest `hal_stm32` module version is correctly pulled.

**Comment by pbeSmartDevice** [comment]
> > I tried to update it, but I can't seem to find the right include and I can't figure out how to include something that is not a file in my `.dts`
> 
> Do you mean this `#include <st/f4/stm32f407v(e-g)tx-pinctrl.dtsi>` ?
> If not working, please try to run `west update` and check that the latest `hal_stm32` module version is correctly pulled.

Yes that was exactly it, everything seems to be working now.

**Comment by pbeSmartDevice** [comment]
@erwango  So is this sufficient?

**Comment by erwango** [comment]
@pbeSmartDevice It seems fine aside from the pinmux part that should be moved to device tree.
I'd advise to wait a few days that SPI and USB are available

**Review by erwango — CHANGES_REQUESTED** [review]
Please squash 2 first commit.
And it would be good to confirm this is working, otherwise there is not point ...

**Comment by pbeTrifork** [comment]
I've Added ADC and SPI as well, should I hold out for USB (I can see it doesn't have a PR yet, at least according to #28999) or should I just squash the three board related ones and have it upstreamed?

Also could I get feedback about the ADC if this is the correct way to go?

**Comment by erwango** [comment]
> Also could I get feedback about the ADC if this is the correct way to go?

Looks good indeed.

> I've Added ADC and SPI as well, should I hold out for USB (I can see it doesn't have a PR yet, at least according to #28999) or should I just squash the three board related ones and have it upstreamed?

USB is on its way, but it might take longer to merge. So let's do the squash and we'll try to merge.

**Comment by pbeTrifork** [comment]
So @erwango I could guess that you would want me to update to get me to use the dts for the USB as well, however I haven't been able to get serial port over USB to work with the new patch and I have no idea why.
The device doesn't come up in `dmesg` when I attach it and the hello-world example therefore can't really work.
blinky works fine, so it's not the setup.

**Comment by erwango** [comment]
@pbeTrifork 

You mean that this is working fine if you configure uart4 pins using pinmux.c ?

You would be able to compare GPIO (A in your case) register settings after pins have been configured (let's say end of `uart_stm32_init`) ?
And if that's equivalent (which I expect to be, since I've verified that several times on various targets), would you be able to look at generated signals in both cases?
I'm thinking this can be a timing issue or a glitch generated on the line. Which would be strange as uart is generally quite error resilient. But I'm not familiar with the probe you're using.

Last point, is there a working branch you can share?

**Comment by pbeTrifork** [comment]
Sorry, It will have to wait as the office had a confirmed COVID-19 case and everyone is staying home (and away from the hardware) until tested.

**Comment by pbeTrifork** [comment]
No, what I mean is that no matter what setting I try out I can't get the hello world sample to work.
When setting the config to 

```
CONFIG_UART_INTERRUPT_DRIVEN=y
CONFIG_UART_CONSOLE_ON_DEV_NAME="CDC_ACM_0"
CONFIG_CONSOLE=y
CONFIG_USB_UART_CONSOLE=y
CONFIG_SERIAL=y
CONFIG_USB=y
CONFIG_USB_DEVICE_STACK=y
CONFIG_USB_DEVICE_OS_DESC=y
```

Then the /dev/ttyACM0 doesn't show up, when I omit fdb6c46 ( changing pinmux settings to dts settings ) it still doesn't show up and dmesg also doesn't show any new connected devices.

**Comment by erwango** [comment]
@pbeTrifork, do you know if there's any life? Is blinky working ?

**Comment by pbeTrifork** [comment]
Yes, blinky is working, my main project for this is also working, I can read the sensor outputs from my program using gdb, but it still doesn't show up on dmesg or as a serial port, so the output doesn't come out.

**Comment by pbeTrifork** [comment]
Merged with master and tried the cdc usb sample, which worked fine.
Seems like I have a problem in my application configuration

File: boards/arm/mikroe_clicker_2/pinmux.c [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+#include <sys/sys_io.h>
+
+#include <pinmux/stm32/pinmux_stm32.h>
+
+/* Pin assignments for MikroE Clicker 2 board for STM32 */
+static const struct pin_config pinconf[] = {
+};
+
+static int pinmux_mikroe_clicker_2_init(const struct device *port)
+{
```
Comment (erwango) [inline boards/arm/mikroe_clicker_2/pinmux.c:20]: This file could be removed now

File: boards/arm/mikroe_clicker_2/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_PINMUX)
+zephyr_library()
+zephyr_library_sources(pinmux.c)
+endif()
```
Comment (erwango) [inline boards/arm/mikroe_clicker_2/CMakeLists.txt:6]: Please remove this file

**Review by erwango — CHANGES_REQUESTED** [review]
Two minor changes and we're good to go, if you're ready.

File: boards/arm/mikroe_clicker_2/pinmux.c (hunk 12 above)
Comment (pbeTrifork) [reply boards/arm/mikroe_clicker_2/pinmux.c:20]: Ok, I thought I read that the idea was to keep it in, but ok.

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
LGTM

**Pull request merged**